lib/src/command.dart

```diff
@@ -46,28 +45,38 @@ final lineLength = stdout.hasTerminal ? stdout.terminalColumns : 80;
 /// of subcommands. Only leaf commands are ever actually invoked. If a command
 /// has subcommands, then one of those must always be chosen.
 abstract class PubCommand extends Command<int> {
-  String get directory => argResults['directory'] ?? _pubTopLevel.directory;
+  @override
+  ArgResults get argResults {
+    final a = super.argResults;
+    if (a == null) {
+      throw StateError(
+          'argResults cannot be used before Command.run is called.');
+    }
+    return a;
+  }
+
+  String? get directory => argResults['directory'] ?? _pubTopLevel!.directory;
 
-  SystemCache get cache => _cache ??= SystemCache(isOffline: isOffline);
+  SystemCache get cache => _cache ??= SystemCache(isOffline: isOffline!);
```

Comment:
Let's try making this a late-final:
```suggestion
  late final SystemCache cache = SystemCache(isOffline: isOffline!);
```